core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -181,21 +178,30 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     return Status.OK;
   }
 
-  private static EquivalentAddressGroup removeDuplicateAddresses(EquivalentAddressGroup eag) {
-    Set<SocketAddress> addressSet = new HashSet<>();
-    ArrayList<SocketAddress> addrs = new ArrayList<>(); // maintains order
-
-    for (SocketAddress address : eag.getAddresses()) {
-      if (addressSet.add(address)) {
-        addrs.add(address);
-      }
+  private static List<EquivalentAddressGroup> deDupAddresses(List<EquivalentAddressGroup> groups) {
+    if (groups == null) {
+      return Collections.emptyList();
     }
 
-    if (addressSet.size() == eag.getAddresses().size()) {
-      return eag;
+    Set<SocketAddress> seenAddresses = new HashSet<>();
+    List<EquivalentAddressGroup> newGroups = new ArrayList<>();
+
+    for (EquivalentAddressGroup group : groups) {
+      if (group == null) {
```

Comment:
This is impossible because we've already validated the groups are not null.

Reply:
removed